File: Code/backend/curriculum/src/main/java/com/scse/curriculum/syllabus/entity/SyllabusVersion.java
package com.scse.curriculum.syllabus.entity;

/** Workflow revisions are whole numbers; there is no minor-revision domain. */
public final class SyllabusVersion {
    private SyllabusVersion() {}

    public static String format(Integer versionNumber) {
        if (versionNumber == null || versionNumber < 1) {
            throw new IllegalArgumentException("Syllabus version number must be a positive integer.");
        }
        return "v" + versionNumber + ".0";
    }

    /** Read legacy labels without mutating historical entities or timestamps. */
    public static String display(Integer versionNumber, String legacyLabel) {
        if (versionNumber != null) {
            return format(versionNumber);
        }
        if (legacyLabel == null || legacyLabel.isBlank()) {
            return null;
        }
        var match = java.util.regex.Pattern.compile("(?i)^(?:Version\\s+|v)?([1-9][0-9]*)(?:\\.0)?$")
                .matcher(legacyLabel.trim());
        if (!match.matches()) {
            throw new IllegalArgumentException("Unsupported syllabus version label: " + legacyLabel);
        }
        return format(Integer.valueOf(match.group(1)));
    }

    public static void requireCanonical(Integer versionNumber, String label) {
        if (!format(versionNumber).equals(label)) {
            throw new IllegalArgumentException("Syllabus version label must be " + format(versionNumber)
                    + "; workflow version labels are system-controlled.");
        }
    }
}

File: Code/frontend/src/lib/syllabusVersion.ts
/** Prefer the backend label; numeric workflow revisions have no minor component. */
export function formatVersionLabel(
  versionNumber?: number | null,
  versionLabel?: string | null,
): string {
  const label = versionLabel?.trim()
  if (label && /^v[1-9]\d*\.0$/.test(label)
      && (versionNumber == null || label === `v${versionNumber}.0`)) {
    return label
  }
  if (versionNumber != null && Number.isSafeInteger(versionNumber) && versionNumber > 0) {
    return `v${versionNumber}.0`
  }
  const legacy = label?.match(/^(?:Version\s+|v)?([1-9]\d*)(?:\.0)?$/i)
  return legacy ? `v${legacy[1]}.0` : "—"
}

/** Numeric ordering, independent of display labels and cohort comparison rules. */
export function compareSyllabusVersions(
  left: { versionNumber?: number | null },
  right: { versionNumber?: number | null },
): number {
  return (left.versionNumber ?? 0) - (right.versionNumber ?? 0)
}
